src/core/tables.js

```diff
@@ -0,0 +1,163 @@
+// @ts-check
+// Module core/tables
+// Handles tables in the document.
+// Generates a Table of Tables wherever there is a #tot element.
+
+import {
+  addId,
+  getIntlData,
+  renameElement,
+  showWarning,
+  wrapInner,
+} from "./utils.js";
+import { html } from "./import-maps.js";
+
+export const name = "core/tables";
+
+const localizationStrings = {
+  en: {
+    list_of_tables: "List of Tables",
+    table: "Table ",
+  },
+  ja: {
+    table: "図 ",
+    list_of_tables: "図のリスト",
+  },
+  ko: {
+    table: "그림 ",
+    list_of_tables: "그림 목록",
+  },
+  nl: {
+    table: "Figuur ",
+    list_of_tables: "Lijst met figuren",
+  },
+  es: {
+    table: "Figura ",
+    list_of_tables: "Lista de Figuras",
+  },
+  zh: {
+    table: "图 ",
+    list_of_tables: "规范中包含的图",
+  },
+  de: {
+    table: "Abbildung",
+    list_of_tables: "Abbildungsverzeichnis",
+  },
+};
+
+const l10n = getIntlData(localizationStrings);
+
+export function run() {
+  const tot = collectTables();
+
+  // Create a Table of Tables if a section with id 'tot' exists.
+  const totElement = document.getElementById("tot");
+  if (tot.length && totElement) {
+    decorateTableOfTables(totElement);
+    totElement.append(
+      html`<h1>${l10n.list_of_tables}</h1>`,
+      html`<ul class="tot">
+        ${tot}
+      </ul>`
+    );
+  }
+}
+
+/**
+ * process all tables
+ */
+function collectTables() {
+  /** @type {HTMLElement[]} */
+  const tot = [];
+  document.querySelectorAll("table").forEach((table, i) => {
+    const caption = table.querySelector("caption");
+
+    if (caption) {
+      decorateTable(table, caption, i);
+      tot.push(getTableOfTablesListItem(table.id, caption));
+    } else {
+      const msg = "Found a `<table>` without a `<caption>`.";
+      showWarning(msg, name, { elements: [table] });
+    }
+  });
+  return tot;
+}
+
+/**
+ * @param {HTMLElement} table
+ * @param {HTMLElement} caption
```

Comment:
Nit:
```suggestion
 * @param {HTMLTableElement} table
 * @param {HTMLTableCaptionElement} caption
```

Reply:
I could use help here. The TypeScript compiler noted that the parameters were not HTMLTableElement.
I could use the TypeScript 'as' keyword to cast from HTMLElement to HTMLTableElement. But these are JavaScript files, not TypeScript.

This code originally comes from the figure code. Apparently figure and figcaption tags are both HTMLElement. There is no HTMLFigureElement for example. So that code didn't warrant an upcast.

Thoughts?

Reply:
I took the liberty to update above in https://github.com/w3c/respec/pull/4200/commits/9175d7528c2e8c83c15e9bc35579451f5147f616